### PR TITLE
Meetings set remote quality

### PR DESCRIPTION
**Pull request by adamweeks**
# Pull Request Template

## Description

This adds the `meeting.setRemoteQualityLevel` function. This will set the quality level in the meeting's mediaProperties object, then calls `updateMedia` to set the max framesize of the offer.

Fixes # https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-133788

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 1]
```diff
@@ -52,20 +55,25 @@ const insertBandwidthLimit = (sdpLines, index) => {
...
+const setMaxFs = (sdp, level = QUALITY_LEVELS.HIGH) => {
+  if (!MAX_FRAMESIZES[level]) {
+    throw new ParameterError(`setMaxFs: unable to set max framesize, value for level "${level}" is not defined`)
+  }
   // eslint-disable-next-line no-warning-comments
   // TODO convert with sdp parser, no munging
   let replaceSdp = sdp;
-  replaceSdp = replaceSdp.replace(/(\na=fmtp:(\d+).*level-asymmetry-allowed=1.*)/gi, `$1;${SDP.MAX_FS_8160}`)
+  const maxFsLine = `${SDP.MAX_FS}${MAX_FRAMESIZES[level]}`
+  replaceSdp = replaceSdp.replace(/(\na=fmtp:(\d+).*level-asymmetry-allowed=1.*)/gi, `$1;${maxFsLine}`)
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:70]: Hopefully we'll get some bandwidth to begin implementing SDP parsers to make areas like this read easier.

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 2]
```diff
@@ -388,38 +401,40 @@ pc.rollBackLocalDescription = (peerConnection) => {
...
- *  peerConnection: {Object} peerConnection
- *  sdpConstraints: {Object} sdp constraints
- * }
- * @param {Object} meetingId meetingId for metrics
+ * @param {Object} params
+ * @param {Object} params.peerConnection
+ * @param {Object} params.sdpConstraints
+ * @param {Object} meetingProperties
+ * @param {string} meetingProperties.meetingId
+ * @param {string} meetingProperties.remoteQualityLevel LOW|MEDIUM|HIGH
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:434]: Good JSDoc readability change.

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 3]
```diff
@@ -388,38 +401,40 @@ pc.rollBackLocalDescription = (peerConnection) => {
...
- * }
- * @param {Object} meetingId meetingId for sending metrics 
- * @returns {Array} [MediaSDP, ScreenSDP]
+ * @param {Boolean} params.offerToReceiveAudio
+ * @param {Boolean} params.offerToReceiveVideo
+ * @param {string} params.offerSdp
+ * @param {MediaStream} params.stream
+ * @param {Object} meetingProperties
+ * @param {string} meetingProperties.meetingId
+ * @param {string} meetingProperties.remoteQualityLevel LOW|MEDIUM|HIGH
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:410]: For situations like these, if we have a Record/Array of constants that could be used in a method, what do we prefer for documenting this?

I'd think:
```js
/**
 * @param {string} meetingProperties.remoteQualityLevel[LOW|MEDIUM|HIGH]
 */
```
or
```js
/**
 * @param {string} ...meetingProperties.remoteQualityLevel
 */
```

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 4]
```diff
@@ -664,6 +664,37 @@ describe('plugin-meetings', () => {
           assert.isRejected(meeting.setLocalVideoQuality('LOW'));
         });
       });
+
+      describe('#setRemoteQualityLevel', () => {
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js:668]: Hurray! Unit Tests!

File: packages/node_modules/samples/browser-plugin-meetings/index.html [hunk 5]
```diff
@@ -309,6 +309,15 @@ <h2>Then click:</h2>
...
         </p>
+        <p>
+          <label for="setRemoteMeetingQualityLevel">Set Remote Quality Level</label>
+          <select id="setRemoteMeetingQualityLevel" name="setRemoteMeetingQualityLevel">
+            <option value="LOW">LOW</option>
+            <option value="MEDIUM">MEDIUM</option>
+            <option value="HIGH">HIGH</option>
+          </select>
+          <button id="setRemoteMeetingQuality" title="meeting.setRemoteQualityLevel()" type="button">meeting.setRemoteQualityLevel()</button>
+        </p>
```
Comment (InteractiveTimmy) [inline packages/node_modules/samples/browser-plugin-meetings/index.html:320]: Good samples coverage!

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 6]
```diff
@@ -1013,3 +1013,9 @@ export const VIDEO_RESOLUTIONS = {
     }
   }
 };
+
+export const MAX_FRAMESIZES = {
+  [QUALITY_LEVELS.LOW]: 1620,
+  [QUALITY_LEVELS.MEDIUM]: 3600,
+  [QUALITY_LEVELS.HIGH]: 8192
+};
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js:1021]: Can we get comments on what these are in reference to quality? Not impacting, but a suggestion.

Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js:1021]: Will add a link to https://en.wikipedia.org/wiki/Advanced_Video_Coding

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 3 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:410]: I actually think a [@typedef](https://jsdoc.app/tags-typedef.html) would be good for this.

**Pull request merged**